### ISSUE TITLE
Configuring Rackspace adapter throws OutOfBoundsException

### ISSUE THREAD

**Issue by jeroenseegers** [1up-lab/OneupFlysystemBundle#86]
Hi,

After following the guide for configuring a Rackspace adapter, I'm presented with an `OutOfBoundsException`:

```
[Symfony\Component\DependencyInjection\Exception\OutOfBoundsException]
  The index "1" is not in the range [0, 0].
```

The way I've configured everything is as follows:

`services.yml`

``` yaml
services:
  my_rackspace_client:
    class: OpenCloud\Rackspace
    arguments:
     - "https://lon.identity.api.rackspacecloud.com/v2.0/"
     - username: "xxx"
       password: "xxx"
```

`config.yml`

``` yaml
oneup_flysystem:
  adapters:
    my_rackspace_adapter:
      rackspace:
        container: my_rackspace_client

  filesystems:
    my_rackspace_filesystem:
      adapter: my_rackspace_adapter
      mount: my_rackspace_filesystem
```

Any clues as to what might be causing this?

**Comment by bytehead** [comment]
Hi Jeroen

Do you have any stacktrace available to this exception?

**Comment by jeroenseegers** [comment]
It seems to have been caused by a missing parameter in the `Resources/config/adapters.xml` file. After adding it this error is resolved, but the bundle still cannot be used because of a second error:

```
Catchable Fatal Error: Argument 1 passed to League\Flysystem\Rackspace\RackspaceAdapter::__construct() must be an instance of OpenCloud\ObjectStore\Resource\Container, instance of OpenCloud\Rackspace given, called in /Users/xxx/Development/xxx/symfony/app/cache/dev/appDevDebugProjectContainer.php on line 2555
```

It appears as though the Rackspace adapter for this bundle has never worked as there's quite some mismatches. I might look into it further, but as I'm under time pressure at the moment I've been forced to switch to the GaufretteBundle which does seem to work out of the box for the Rackspace adapter.

**Comment by bytehead** [comment]
Thanks for investigating! We will try to fix this in a next release.

**Comment by mediamonks-robert** [comment]
Also running into this issue, any fix available already?

**Comment by guiguidu31300** [comment]
@mediamonks-robert : Gaufrette project is half dead (many merge pull, many issues and not many commit) for a while. So we switch from Gaufrette to Flysystem (also for the temporary URL generator)

Here is my setup works

config.yml :
```yml
services:
    opencloud_connection:
        lazy: true
        class: OpenCloud\OpenStack
        arguments:
            - "https://auth.cloud.ovh.net/v2.0/"
            - {username: %openstack_username%, password: %openstack_password%, tenantName: "xxxxxxxx"}
            - {logger: @monolog.logger}
    opencloud_client:
        lazy: true
        class: OpenCloud\ObjectStore\Service
        factory: [@opencloud_connection, objectStoreService]
        arguments:
            - 'swift'
            - 'SBG1'
            - 'publicURL'
    opencloud_document_container:
        lazy: true
        class: OpenCloud\ObjectStore\Resource\Container
        factory: [@opencloud_client, getContainer]
        arguments:
            - "documents-prod"
oneup_flysystem:
    adapters:
        document:
            rackspace:
                container: opencloud_document_container
                lazy: true
    filesystems:
        documents:
            adapter: document
            mount: documents
```

**Comment by mediamonks-robert** [comment]
@guiguidu31300 Thanks a lot for the information and the solution! I'll stick to flysystem then :)

**Comment by mediamonks-robert** [comment]
I managed to create a rackspace document container with this config:

```
parameters:
    rackspace_cloudfiles_url: https://identity.api.rackspacecloud.com/v2.0/
    rackspace_cloudfiles_region: DFW
    rackspace_cloudfiles_username: my_username
    rackspace_cloudfiles_apikey: my_apikey
    rackspace_cloudfiles_bucket: my_bucket

services:
    rackspace_connection:
        lazy: true
        class: OpenCloud\Rackspace
        arguments:
            - %rackspace_cloudfiles_url%
            - {username: %rackspace_cloudfiles_username%, apiKey: %rackspace_cloudfiles_apikey%}
    rackspace_client:
        lazy: true
        class: OpenCloud\ObjectStore\Service
        factory: [@rackspace_connection, objectStoreService]
        arguments:
            - 'cloudFiles'
            - %rackspace_cloudfiles_region%
            - 'publicURL'
    rackspace_document_container:
        lazy: true
        class: OpenCloud\ObjectStore\Resource\Container
        factory: [@rackspace_client, getContainer]
        arguments:
            - %rackspace_cloudfiles_bucket%
```
And then I can pass "rackspace_document_container" as the container argument to this bundle.

However after this it was still failing. I did some digging and the problem is with the version I'm using (1.2.0) where the service definition of the rackspace adapter does not have the "prefix" option. Since the service builder is assuming it's there it will try to replace a non existing argument and will fail with the above error. It seems this is fixed in the current master (and then also the "lazy" option is implemented) so I'll switch to that for now and hope it works.

**Comment by twify93** [comment]
OK, guys help me with this configuration, because I am fighting from couple of days now...
I tried to fix it, but then I realized there is no way to pass the required Container I want to use!

> Catchable Fatal Error: Argument 1 passed to OpenCloud\ObjectStore\Resource\Container::__construct() must be an instance of OpenCloud\Common\Service\ServiceInterface, string given

```
# Rackspace services
services:
    rackspace_connection:
        lazy: true
        class: OpenCloud\Rackspace
        arguments:
        - %rackspace_cloudfiles.url%
        - {username: %rackspace_cloudfiles.username%, apiKey: %rackspace_cloudfiles.apikey%}
    rackspace_client:
        lazy: true
        class: OpenCloud\ObjectStore\Service
        factory: [@rackspace_connection, objectStoreService]
        arguments:
        - 'cloudFiles'
        - %rackspace_cloudfiles.region%
        - 'publicURL'
    rackspace_document_container:
        lazy: true
        class: OpenCloud\ObjectStore\Resource\Container
        factory: [@rackspace_client, getContainer]
        arguments:
        - %rackspace_cloudfiles.bucket%
# FlySystem filesystems
oneup_flysystem:
    adapters:
        booking_document_adapter:
            rackspace:
                container: rackspace_document_container
                prefix: %oneup_flysystem.main_upload_folder%/%upload.booking_documents_folder%
```
I end up doing this:
```
# Rackspace services
services:
      rackspace_connection:
          lazy: true
          class: OpenCloud\Rackspace
          arguments:
              - %rackspace_cloudfiles.url%
            - {username: %rackspace_cloudfiles.username%, apiKey: %rackspace_cloudfiles.apikey%}
    rackspace_client:
        lazy: true
        class: OpenCloud\ObjectStore\Service
        factory: [@rackspace_connection, objectStoreService]
        arguments:
            - @rackspace_connection
            -  null
            - 'cloudFiles' # this should be cloudFiles
            - %rackspace_cloudfiles.region% # the region for Fulcrum is LON
            - 'internalURL' # this should be internalURL
    rackspace_container:
        lazy: true
        class: OpenCloud\ObjectStore\Resource\Container
        factory: [@rackspace_client, getContainer]
        arguments:
            - @rackspace_client
# FlySystem filesystems
oneup_flysystem:
    adapters:
        booking_document_adapter:
            rackspace:
                container: rackspace_container
                prefix: %oneup_flysystem.main_upload_folder%/%upload.booking_documents_folder%
```

In both configurations provided by @mediamonks-robert and @guiguidu31300 in 
```
rackspace_document_container:
....
    arguments:
        - %rackspace_cloudfiles.bucket%
```
there is a "bucket" provided as argument to the Container class, but when I open the class the constructor looks like this:

```
__construct(ServiceInterface $service, $data = null)
```

No Container/Bucket specified.

**Comment by mediamonks-robert** [comment]
The "rackspace_document_container" service is created by calling getContainer() on "rackspace_client" service. That getContainer() method has a $data parameter which is the bucket name.

**Comment by twify93** [comment]
Yes, but the strange thing is that it's not trying to call getContainer, but rather is trying to make new instance using the constructor?

**Comment by mediamonks-robert** [comment]
Maybe you need to empty your cache? The setup is still working fine for me on Symfony 2.8.

**Comment by bytehead** [comment]
Fixed in #89.

**Comment by maniekx1984** [comment]
@guillaume-ro-fr - your solution is still cool for my app :)
But, can you advice me something?

Do you know how to configure ACL? I'd like to have private bucket but with public-read files (only files, without listing directories) - did you configure it in your app?
Second - public links, have you any idea how to use short links, not e.g. https://storage.waw1.cloud.ovh.net/v1/AUTH_SOMETEXT/BUCKET but e.g. https://storage.waw1.cloud.ovh.net/BUCKET/directory/file.jpg
?
